### PR TITLE
New parser for observations used by default with the possibility of opting out through an environment variable

### LINKED ISSUE #5367: Lark parser for observation configuration

### PR DESCRIPTION
**Pull request by eivindjahren**
**Issue**
Resolves #5367

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Updated documentation
- [x] Ensured new behaviour is tested

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: src/ert/parsing/new_observations_parser.py [hunk 1]
```diff
@@ -48,21 +37,420 @@ def from_rule(cls, rule: str) -> "ObservationType":
...
+    error_mode = "ABS"
+    summary_key = None
+
+    date_dict: DateDict = {}
+    float_values: Dict[str, float] = {"ERROR_MIN": 0.1}
+    for key, value in inp.items():
+        if key == "RESTART":
+            date_dict["RESTART"] = validate_positive_int(value, key)
+        elif key in ["ERROR", "ERROR_MIN"]:
+            float_values[str(key)] = valdiate_positive_float(value, key)
```
Comment (oysteoh) [inline src/ert/parsing/new_observations_parser.py:231]: valdiate -> validate

File: src/ert/parsing/new_observations_parser.py [hunk 2]
```diff
@@ -48,21 +37,420 @@ def from_rule(cls, rule: str) -> "ObservationType":
...
+        elif key in ["DAYS", "HOURS"]:
+            date_dict[str(key)] = valdiate_positive_float(value, key)
```
Comment (oysteoh) [inline src/ert/parsing/new_observations_parser.py:233]: valdiate -> validate

File: src/ert/_c_wrappers/enkf/enkf_obs.py [hunk 3]
```diff
@@ -26,18 +43,61 @@
...
+        return name in self.values
+
+    def __getitem__(self, name: str) -> Any:
+        return self.values[name]
+
+
+class _HistoryConfInstance(_DictConfInstance):
+    def __init__(self, name: str, values: HistoryValues):
+        self.name = name
+        self.values = values
```
Comment (DanSava) [inline src/ert/_c_wrappers/enkf/enkf_obs.py:70]: Why not just `super().__init(name, values)` ? in place of the explicit member assignment?

File: src/ert/parsing/new_observations_parser.py (hunk 1 above)
Comment (eivindjahren) [reply src/ert/parsing/new_observations_parser.py]: fixed

File: src/ert/parsing/new_observations_parser.py (hunk 2 above)
Comment (eivindjahren) [reply src/ert/parsing/new_observations_parser.py]: fixed

File: src/ert/_c_wrappers/enkf/enkf_obs.py (hunk 3 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/enkf/enkf_obs.py]: I updated it to use the inherited constructor.

File: src/ert/_c_wrappers/enkf/enkf_obs.py [hunk 4]
```diff
@@ -26,18 +42,57 @@
...
+    def get_sub_instances(self, name: Literal["SEGMENT"]):
+        segments = self.values[name]
+        return [_DictConfInstance(segment[0], segment[1]) for segment in segments]
+
+
+class _NewParserAdapter:
+    def __init__(self, conf_content: ConfContent):
+        self.conf_content = conf_content
+
+    def get_sub_instances(self, name: str):
```
Comment (DanSava) [inline src/ert/_c_wrappers/enkf/enkf_obs.py:76]: Why not just break these down on initialization like 
```python 
self.history_observations =[]
self.general_observations =[]
self.summary_observations =[]

for _type, name, values, _ in self.conf_content: # ', _' required if there are more elements in the sub lists of self.conf_content
  if _type == ObservationType.HISTORY:
     self.history_observations.append(_HistoryConfInstance(name, values))
  if _type == ObservationType.GENERAL:
     self.general_observations.append(_DictConfInstance(name, values)) 
  if _type == ObservationType.SUMMARY:
     self.summary_observations.append(_DictConfInstance(name, values)) 
```

Ignore the above suggestion I realized `get_sub_instances` is required for compatibility reasons with other parts

**Review by DanSava — APPROVED** [review]
Looks very good, nice one! We might need to look into refactoring some of the names in enkf_obs.py, there is a lot of `instance` all over the place conf_instance, sub_instance, instance of sub_instance, and it is a bit confusing at times what each is actually referring to.

File: src/ert/_c_wrappers/enkf/enkf_obs.py (hunk 4 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/enkf/enkf_obs.py:76]: I think we can do something even simpler once the old parser is gone, and not sure if I can get the old parsers output to be iterated over like you suggested.

**Comment by eivindjahren** [comment]
> Looks very good, nice one! We might need to look into refactoring some of the names in enkf_obs.py, there is a lot of `instance` all over the place conf_instance, sub_instance, instance of sub_instance, and it is a bit confusing at times what each is actually referring to.

I think we will remove most of the `Instance` names with the old parser.

File: src/ert/parsing/new_observations_parser.py [hunk 5]
```diff
@@ -48,21 +37,420 @@ def from_rule(cls, rule: str) -> "ObservationType":
...
+        elif key in ["DAYS", "HOURS"]:
+            date_dict[str(key)] = validate_positive_float(value, key)
+        elif key == "VALUE":
+            float_values["VALUE"] = validate_float(value, key)
+        elif key == "ERROR_MODE":
+            error_mode = validate_error_mode(value)
+        elif key == "KEY":
+            summary_key = value
+        elif key == "DATE":
+            date_dict["DATE"] = value
```
Comment (andreas-el) [inline src/ert/parsing/new_observations_parser.py:241]: Replace lookup strings with key? There are some other lookups that could be changed elsewhere as well.

File: src/ert/_c_wrappers/enkf/enkf_obs.py [hunk 6]
```diff
@@ -36,18 +42,57 @@
 
 DEFAULT_TIME_DELTA = timedelta(seconds=30)
 
+USE_NEW_PARSER_BY_DEFAULT = True
```
Comment (andreas-el) [inline src/ert/_c_wrappers/enkf/enkf_obs.py:45]: Use unique name for this? This is the same as the original lark implementation.

**Review by andreas-el — APPROVED** [review]
Minor comments, looks good!

File: src/ert/parsing/new_observations_parser.py (hunk 5 above)
Comment (eivindjahren) [reply src/ert/parsing/new_observations_parser.py]: Should be fixed now.

File: src/ert/_c_wrappers/enkf/enkf_obs.py (hunk 6 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/enkf/enkf_obs.py]: Fixed

**Pull request merged**